File: EPI/Recursion/Compute a Gray code/swift/16.10.swift
public func grayCode(bitCount: Int) -> [Int] {
    if bitCount <= 0 {
        return [0]
    }

    let prev = grayCode(bitCount: bitCount -  1)
    let leadingOne = 1 << (bitCount - 1)

    return prev + prev.reversed().map { leadingOne + $0 }
}

File: EPI/Recursion/Compute a Gray code/typescript/16.10.ts
export default function grayCode(bits: number): number[] {
    if (bits === 0) {
        return [0];
    }

    const prev = grayCode(bits - 1);

    return [
        ...prev,
        ...prev.reverse().map(v => v | (1 << (bits - 1))),
    ];
}
